Add support for 24-bit color

u-quark opened the pull request:
References for 24-bit color support in the terminal:
https://github.com/termstandard/colors
https://gist.github.com/XVilka/8346728 (older version of the above)

Closes #167.

ikea-shark-official committed:
minor layout fixes

ikea-shark-official committed:
add an option to configure 24bit output

create a datatype `ColorMode`, which is stored in `Config`
and `Output`.  It is passed to all of the functions that handle color
processing/output.

Currently this only has one possible value, which gives us the default
vty color handling behavior.

Factor out color drawing code in TerminfoBased.hs

Add datatype for 24-bit colors

Add 24-bit color support

we just write the colors straight to the terminal without using
terminfo.

Move all color-handling info to `ColorMode`

- Add terms for 1/8/16/240 colors to `ColorMode`
- Add function `detectColorMode` to detect color mode

This removes all color-handling information outside of the `ColorMode`
datatype, giving us a single source for color-support information.

This emulates the old color-handling code perfectly pretty much perfectly,
with one exception; For users with >256 colors listed, we don't do anything
special when outputting `Color240`s. The old behavior was absolutely
baffling to me, and seemed like it might cause future issues, so I just
removed it.

Add functions for creating 24-bit colors

- linearColor is just a polymorphic synonym for `RGBColor`
- srgbColor converts inputs from to linear before making a `RGBColor`
- color240 is the same as rgbColor, but with a less confusing name

This is all based of of the assumption that outputs will be linear,
which is true for gnome-terminal, but I haven't tested it with anything
else

u-quark committed:
Read COLORTERM env to determine truecolor support

The environment variable COLORTERM is used by terminal emulators to
signify that they support truecolor in color escape sequences.

u-quark committed:
Fix tests for full color support

jtdaugherty commented on src/Graphics/Vty/Attributes/Color.hs:
```diff
@@ -125,12 +126,15 @@ detectColorMode termName' = do
+        | termColors <  16              -> ColorMode8
+        | termColors == 16              -> ColorMode16
+        | termColors <  256             -> ColorMode240 (fromIntegral termColors - 16)
+        | colorterm == Just "truecolor" -> FullColor
```
These checks assume that the value of `termColors` obtained from `terminfo` will always be larger than 256 if the environment variable is set. Is that a safe assumption?

u-quark replied:
I think it is a safe assumption. The value `termColors` is based on `TERM` environment variable. See the testing I did for more details. For most terminals it seems `termColors == 256` and they support at least `ColorMode240`. It seems some terminals do not advertise their full capabilities with the environment variables `TERM` and `COLORTERM`. I am happy to switch the order of the checks. I think the code as it is makes more sense though, especially in the case when `COLORTERM` is set for whatever reason and not overridden by the terminal emulator.

jtdaugherty replied:
Okay, thanks for the extra details.

jtdaugherty commented on src/Graphics/Vty/Attributes/Color.hs:
```diff
@@ -119,9 +157,40 @@ brightMagenta= ISOColor 13
+        r' = fromIntegral r :: Word8
+        g' = fromIntegral g :: Word8
+        b' = fromIntegral b :: Word8
+        makeError = error $ printf "invalid value in rgbColor: (%d, %d, %d)" r' g' b'
```
I don't know if you have the time to make a change to this PR -- I know it's been a while since you wrote it -- but if so, I think it would be better if this clamped color values to the [0..255] range rather than raising an exception.

u-quark replied:
Yeah, it is better. Fixed.

u-quark committed:
Add 24 bit color demo

u-quark committed:
Clamp values to [0..255] instead of error

u-quark commented:
I rebased the code on top of current master with minor changes.

For references I am using these:
https://github.com/termstandard/colors
https://gist.github.com/XVilka/8346728 (older version of the above)

I added a full 24bit color block in the vty-demo.

I also tested a few terminals:

- GNOME Terminal 3.40.3 using VTE 0.64.2 +BIDI +GNUTLS +ICU +SYSTEMD
  TERM=xterm-256color COLORTERM=truecolor termColors=256
  ColorMode240 -> works, FullColor -> works
- Konsole 21.08.1
  TERM=xterm-256color COLORTERM=truecolor termColors=256
  ColorMode240 -> works, FullColor -> works
- Kitty 0.19.3
  TERM=xterm-kitty COLORTERM=truecolor termColors=256
  ColorMode240 -> works, FullColor -> works
- mlterm 3.9.0
  TERM=mlterm COLORTERM is unset termColors=8
  ColorMode240 -> no color, FullColor -> no color
  though with explicitly set TERM=mlterm-256color and COLORTERM=truecolor: termColors=256
  ColorMode240 -> works, FullColor -> works with some color quantitation
- urxvt (rxvt-unicode) v9.22 - released: 2016-01-23
  TERM=rxvt-unicode-256color COLORTERM=rxvt-xpm termColors=256
  ColorMode240 -> works, FullColor -> falls back to ColorMode240 mode
  It seems urxvt does not support truecolor in this version.
- XTerm 368
  TERM=xterm COLORTERM is unset termColors=8
  ColorMode240 -> no color, FullColor -> no color
  though with explicitly set TERM=xterm-256color and COLORTERM=truecolor: termColors=256
  ColorMode240 -> works, FullColor -> works
- st 0.8.4
  TERM=st-256color COLORTERM is unset termColors=256
  ColorMode240 -> works, FullColor -> falls back to ColorMode240 mode
  though with explicitly set COLORTERM=truecolor: termColors=256
  ColorMode240 -> works, FullColor -> works
- tmux 3.2a
  TERM=tmux-256color COLORTERM=truecolor termColors=256
  ColorMode240 -> works, FullColor -> works

All these with libtinfo6/ncurses version 6.2

On Mac:
- iTerm2
  TERM=xterm-256color COLORTERM=truecolor termColors=256
  Did not test with vty-demo
- Terminal.app
  TERM=xterm-256color COLORTERM=truecolor termColors=256
  Did not test with vty-demo

jtdaugherty commented:
@u-quark Thank you so much for your work on this and for following up recently after I let it sit for so long!

u-quark commented:
Thanks for merging this @jtdaugherty!
